pyqtgraph/colorRegistry.py

```diff
@@ -0,0 +1,354 @@
+]:
+    DEFAULT_COLORS[key] = DEFAULT_COLORS[col]
+
+del key, col # let the linter know that we are done with these
+
+def _expand_rgba_hex_string(hex):
+    """ normalize hex string formats and extract alpha when present """
+    length = len(hex)
+    if length < 4   : return None, None
+    if hex[0] != '#': return None, None
+    if length == 4:
+        alpha = 255
+        return '#' + hex[1]*2 + hex[2]*2 + hex[3]*2, alpha
+    if length == 5:
+        alpha = int(2*hex[4], 16)
+        return '#' + hex[1]*2 + hex[2]*2 + hex[3]*2, alpha
+    if length == 7:
+        alpha = 255
+        return '#'+hex[1:7], alpha
+    if length == 9:
+        alpha = int(2*hex[7:9], 16)
+        return '#'+hex[1:7], alpha
+    else:
+        return None, None
+
+# An instantiated QObject is required to emit QSignals. 
+# functions.py initializes and maintains COLOR_REGISTRY for this purpose.
+class ColorRegistry(QtCore.QObject):
+    """
+    Provides palette change signals and maintains color name dictionary
+    Typically instantiated by functions.py as COLOR_REGISTRY
+    Instantiated by 'functions.py' and retrievable as functions.COLOR_REGISTRY
+    """
+    paletteHasChangedSignal = QtCore.Signal() # equated to pyqtSignal in qt.py for PyQt
+    _registrationGenerator = itertools.count()
+
+    def __init__(self, color_dic):
+        """ initialization """
+        super().__init__()
+        self.color_dic = color_dic # this is the imported functions.Colors!
+        self.color_dic.clear()
+        self.color_dic.update( DEFAULT_COLORS)
+        # set of objects that are registered for update on palette change:
+        self.registered_objects = {} # stores registration: (weakref(object), descriptor), deletion handled by weakref.finalize
+        self.color_cache = weakref.WeakValueDictionary() # keeps a cache of QColor objects, key: (name, alpha)
+        self.pen_cache   = weakref.WeakValueDictionary() # keeps a cache of QPen   objects, key: (name, width, alpha)
+        self.brush_cache = weakref.WeakValueDictionary() # keeps a cache of QBrush objects, key: (name, alpha)
+
+    def confirmColorDescriptor(self, color):
+        """
+        Takes a color or brush argument and tries to convert it to a normalized descriptor in the form
+        ``( str(name), None/int(alpha) )``. Returns False if this is not possible.
+        """
+        name = alpha = None
+        if color is None or isinstance( color, str):
+            name = color
+        elif isinstance( color, (collections.abc.Sequence, np.ndarray) ): # list-like, not a dictionary
+            length = len(color)
+            if length == 0: 
+                return False
+            if length >= 1:
+                name  = color[0]
+                # if color id is list-like, the first element needs to be str or None
+                if not ( color is None or isinstance(name,str) ): return False
+            if length >= 2: alpha = color[1]
+        else:
+            return False
```

Comment:
do we want to return False, or should we raise some kind of exception?  I always find it odd when a function returns False, but not True (or True and not False).  An alternative would be to return None, an empty tuple (which doesn't feel right to me either).

Reply:
This is in `colorRegistry.confirmColorDescriptor()`, which is tasked with checking that a given set of arguments can function as a 'descriptor' as used specifically by the color registry.
There are formats that are valid parameters for `mkColor()`, but not for the registry: Like `(255,0,0)` for red. Right now `mkColor()` just passes that to `getRegisteredColor()` and thus to `confirmColorDescriptor()`. `False` is a perfectly valid reply: It just means that the color registry will not handle this, since it looks for descriptors like 'r' or ('r',100). It does (reluctantly) handle the hex codes now ('#FF0000'), but only because all the required parsing is already done, and it seems wasteful to bounce back to `mkColor()` just repeat the same code...

If we decide to raise an error (probably in `getRegisteredColor()` instead of here) for invalid inputs, then we need a try/except block in `mkColor()`  

Is that acceptable?